### DISCUSSION TITLE
Creation and manipulation of camera objects

### DISCUSSION THREAD

**Discussion by perkalkyl** [nxrighthere/UnrealCLR#43] [Help] [answered]
Hi! I am kinda ashamed to have spent a couple of hours on this, and I just don't get it.
I am totally new to UE, and the whole Content/PlayerControl/Camera/etc logic is not clear to me.
If someone would reach out and help me, I would be superhappy :-D .

I have created a Blueprint class of typ CameraActor in my content named "bpCameraActor". I have set the properties as I want it to be. I now want to do the following at the creation of my Testclass (that gets instantiated at Main.OnWorldBegin )

Problem 1:
1. - Add 2 instances of bpCameraActor to the scene in code.
2. - Switch the active view to instance 1 or 2 in code
3. - Change the location and/or rotation of the selected instance at OnTick() in code

Problem 2:
If I create a pawn object with a camera attached to it and place it in a scene, how do i select that camera in code to be the active view? I have understood how to select a named object, like a cube named "myCube", but if it has a camera attached to it, how do I reach that object?

To be clear: The C# part of creating classes, instantiation, Interfaces etc. is not a problem for me. It's the creation of, and required class structure to UE objects that is a puzzle to me. Can I move a camera in code, or has it have to be somehow related to a player component? And even the words in that question is mostly a guess :-) 

Help please :-)

BR
Per

**Comment by nxrighthere** [comment, marked as answer]
First things first, make sure that you are familiar with [GameMode](https://docs.unrealengine.com/en-US/Gameplay/Framework/GameMode/index.html) and [how it's used](https://www.youtube.com/watch?v=kiBQ4Dsj34w). This API is not exposed yet, but for now, you can use a blueprint to set up default classes with `PlayerController` and set a custom `PlayerCameraManager`.

> Problem 1

That said, to just switch between cameras, you can use the [same](https://www.youtube.com/watch?v=KpSzDxMLoHU) approach but from managed code with [`PlayerController.SetViewTarget()`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/PlayerController-SetViewTarget(Actor).md) or [`PlayerController.SetViewTargetWithBlend()`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/PlayerController-SetViewTargetWithBlend(Actor_float_float_BlendType_bool).md). To change the location retrieve `SceneComponent` from the `Camera` actor.

> Problem 2

The most straightforward way is to retrieve an actor from the world by predefined [name](https://github.com/nxrighthere/UnrealCLR/blob/master/API/World-GetActor-T-(string).md) or [tag](https://github.com/nxrighthere/UnrealCLR/blob/master/API/World-GetActorByTag-T-(string).md), and get an attached component of a specified type. This functionality is based on generics, so you only need to specify the appropriate types, optionally with specified names or tags.

> To be clear: The C# part of creating classes, instantiation, Interfaces etc. is not a problem for me.

Please, use learning materials from the internet if it's not related to the plugin itself. It will take less time to get things done rather than wait for a reply here, and most of the C++ and blueprints experiences can be used to write managed code with the plugin.

**Comment by perkalkyl** [comment]
Dear Stanislav

Please bare with me :-) I have spent hours trying to get a C++ example that makes sense, but failed. I truly am not wasting your time without first trying to solve my issues by other means. 
With that said:
I am actually getting forward in my tests :-) BUT! And this drives me crazy.. If I instantiate a pawn or an actor of any kind, like a simple Cube in C#, I can't get it to show up in the model. What am I missing? Does it need to be attached to some other object somehow? I have built my class based on your InstancedStaticMeches as your rotating cubes obviously are visible :-D and I know my objects are there and alive. I can set my ViewTarget to the pawn and move it around and that works fine. But they don't show up visually. I can drag the same content object out on the Level in the UE editor, and it shows up in all it's glory (well.... its a gray box, but still!)

So basically the code to create an instance of a content is:

```
private Actor cube;
private SceneComponent SC_cube;


public void OnBeginPlay() {

        PlayerController pc = World.GetFirstPlayerController();
	cube = new Actor("myTestKub", Blueprint.Load("Game/KKActor"));
	SC_cube = new(cube);
	SC_cube.SetWorldLocation(new(0, 0, 40.0f));
	pc.SetViewTarget(World.GetActor<Camera>("Camera1"));

}

public void OnTick(float deltaTime){
  	SC_cube.AddWorldOffset(new Vector3(0,1,0));
        SC_cube.UpdateToWorld();
}

```

Can you please help?

BR
Per

**Reply by nxrighthere** [reply]
> If I instantiate a pawn or an actor of any kind, like a simple Cube in C#, I can't get it to show up in the model. What am I missing? Does it need to be attached to some other object somehow?

You are using a simple scene component that can be transformed or attached but has no rendering or collision capabilities, as stated in the [API reference](https://github.com/nxrighthere/UnrealCLR/blob/master/API/SceneComponent.md). To render basic objects, you should use classes derived from the [`PrimitiveComponent`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/PrimitiveComponent.md) and attach it to your actor. See [static meshes test](https://github.com/nxrighthere/UnrealCLR/blob/master/Source/Managed/Tests/StaticMeshes.cs).
